README
$Id: README,v 1.20 2008-02-04 18:02:44 mike Exp $


Introduction
------------

This directory contains the source code for Index Data's open source
link resolver, Keystone Resolver, which is part of the Keystone
Digital Library suite.  It is implemented as a Perl module called
"Keystone::Resolver".

TROUT was our earlier proof-of-concept implementation of a trivial
OpenURL resolver: its name stood for Trout Resolves Open URLs
Trivially.  The code was trivial because it was based on a trivial
standard: OpenURL v0.1, as described in the ten-page document
	http://www.openurl.info/registry/docs/pdf/openurl-01.pdf

The new code does not have this luxury for three reasons:

    1.	It is not limited to resolving OpenURLs, but also intends to
	handle DOIs and, in principle at least, other forms of
	metadata-based link.

    2.	Its OpenURL support is based on the newer and much more
	verbose version of the standard as produced by ANSI/NISO
	Committee AX and described at
		http://library.caltech.edu/openurl/Standard.htm
	This standard abstracts and indirects absolutely everything,
	whether it needs abstracting or not, and the code needs to
	reflect this.

    3.	Unlike TROUT, Keystone Resolver needs to do non-trivial things
	in order to resolve links: in particular, it needs a big,
	complex knowledge-base that tells it what resources are
	available to link to and what they contain.

Accordingly, the new code comes in lots of classes, which are
described in the file "Classes".  If you are about to read the
resolver code, that file is a good place to start.


Public CVS Download
-------------------

 cvs -d :pserver:cvs@bagel.indexdata.dk:/cvs login

 use password 'anonymous'

 cvs -d :pserver:cvs@bagel.indexdata.dk:/cvs co openurl-resolver 


Directory Structure
-------------------

The Keystone Resolver distribution is laid out in the following
directories:

bin/	Resolver-related scripts to be run from the command-line.

db/	Resource database material, including schemas, sample data and
	database-creation utilities.  At present, this is set up to
	make a tiny "toy" database.  In future releases, it will be
	expanded to make further databases, including one based on
	CUFTS data.

doc/	Embryonic documentation, in plain text format.  Eventually
	this will either be moved into Perl POD format (in the "lib"
	directory with the source code) or formatted using a proper
	system such as DocBook or OpenOffice.

etc/	Various configuration files, including XML DTDs and XSLT
	stylesheets.

lib/	The resolver source-code library.  (The actual resolver
	program is a trivial seven-line script in the
	web/htdocs/mod_perl/ area -- the library does all the work.)

t/	Test scripts, invoked by the distribution's "make test" rule.
	See also the t/regression subdirectory and its README file.

web/	The resolver's web-server files: server configuration files,
	CGI/mod_perl scripts, HTML pages, images, stylesheets ...

The purpose and contents of most of these directories are described in
more detail in their own README files.

If you got this software via CVS rather than as a distribution
tarball, then you will also have an "archive" directory.  The whole
purpose of this is to contain all the stuff that's not interesting to
anyone except the developers, so just delete it :-)


Prerequisites
-------------

-> A web-server.
Any web server that supports the CGI standard should work, but we use
Apache 1.3 and Apache 2.0 with mod_perl.  The rest of these
instructions assume that's what you're using, and the Debian packaging
includes support for these servers.

-> The Perl module CGI
This is not used by the main resolver entry point, but by the
utility method Keystone::Resolver::OpenURL->newFromCGI(), which uses
it to gather the arguments to pass into the Resolver library proper.
So in theory at least we can use the same library to make resolvers
that get their arguments some other way, e.g. link resolution by
email.

-> The Perl module DBI
This is used to access the resource database.  You also need the Perl
module forwhatever driver you use, e.g. DBD::MySQL.

-> The actual database software, e.g. MySQL
You should be able to use any relational database (MySQL,
PostgreSQL, Oracle, etc.), but the development has been done
using MySQL and it'll be simpler to use that unless you have a
compelling reason to do something different.

-> The Perl module LWP
This is used to resolve the enormous number of network indirections
that a v1.0 OpenURL can have, e.g. the OpenURL itself can use a
By-Reference transport, the ContextObject can specify any or all of
the six entities by reference.

-> The Perl module XML::LibXSLT
This is used to transform the resolver's XML output into pretty,
user-facing HTML.
	-> Gnome libxslt, including development kit
	-> The Perl module XML::LibXML
		-> Gnome libxml2, including development kit
		-> The Perl module XML::SAX
		-> The Perl module XML::NamespaceSupport
		-> The Perl module XML::LibXML::Common

-> The Perl module Text::Iconv
This is used to translate between different character encodings.
	-> The iconv library, but this seems to be included in libc
	   (the standard C library) in Red Hat 9, and therefore
	   probably also in most modern operating systems.

-> The Perl module Digest::MD5
This is needed to calculate the checksums that Elsevier requires in
the customer-specific URLs that access its full-text documents.

-> The Perl module HTML::Mason
This is needed to power the admin pages.


Installation
------------

To install this module type the following:

	perl Makefile.PL
	make
	make test
	sudo make install

You will also need to build the "toy" resource database (or of course
a proper one if you have the data).  To do this, run "make" in the
"db" subdirectory, providing the root MySQL password when requested to
do so.  This will allow the bin/kr-test and
web/htdocs/mod_perl/resolve scripts to run successfully.

Once the toy database has been built, it's possible to run a simple
sanity-test without installing or even building anything, using the
kr-test script:

	perl -I lib bin/kr-test t/regression/zetoc-suuwassea


Installation the Debian way
---------------------------

Building Debian packages

        perl Makefile.PL
        dpkg-buildpackage -rfakeroot
        
        cd ../
        sudo dpkg -i libkeystone-resolver-perl_1.15-1_all.deb
        sudo dpkg -i keystone-resolver_1.15-1_all.deb


Configuration
-------------

To set up Keystone Resolver, you need to do the following steps:

* If you're going to run the resolver as a virtual host (which is what
  I do), create an entry in /etc/hosts for the hostname, for example
  x.resolver.indexdata.com -- or of course set up DNS to serve that
  name's IP address.

  Notice that the example MySQL RDB only includes the service names 
  'id', 'dbc', 'talis', 'resolver', and 'localhost', and your virtual host 
  name needs to use one of those if you do not add another ressource name 
  in the RDB.


* Configure your web server so it can execute the resolver code.  If
  you're using Apache 1.3, you can use a lightly tweaked copy of the
  sample configuration file
	web/conf/apache1.3/xeno.conf
  from this distribution.  Just drop it into the server's
  configuration directory, usually /etc/httpd/conf.d or something
  similar depending on what operating system you're using.  Note that
  you will in general need to change the hostnames in this file.

Non-standard installation directory
-----------------------------------

This software expects to be unpacked into the directory
	/usr/local/src/cvs/resolver/
That path is wired into several places.  If you want to run it from
somewhere else, you'll need to change them all:

* The DocumentRoot, Directory, PerlSetEnv and Alias directives in
  web/conf/apache1.3/xeno.conf (or whatever Apache configuration
  you're using)
* The "xsltdir" setting in lib/Keystone/Resolver.pm

Clearly this is too many places; we should try to find a way to reduce
it, ideally to a single place.


Support
-------

Informal support is available on the Keystone Resolver community
mailing list at
	http://www.indexdata.dk/mailman/listinfo/resolver
which any user is free to join.

Commercial support is available from Index Data.
Email <info@indexdata.com> for details.


Copyright and Licence
---------------------

Copyright (C) 2004-2007 Index Data Aps.

This library is free-as-in-freedom software (which means it's also
open source); it is distributed under the GNU General Public Licence,
version 2.0, which allows you every freedom in your use of this
software except those that involve limiting the freedom of others.
A copy of this licence is in the file "GPL-2"; it is described and
discussed in detail at
	http://www.gnu.org/copyleft/gpl.html

The primary author is Mike Taylor <mike@indexdata.com>



Changes
$Id: Changes,v 1.75 2008-04-30 17:27:56 mike Exp $

Revision history for Perl extension Keystone::Resolver.

1.22  (IN PROGRESS)
	- Oracle back-end now fully functional and tested, both with
	  the resolver itself and the Admin UI.  See db/README for
	  instructions on making a toy Oracle resource database, and
	  the top-level README for instructions on running the
	  test-suite against this.
	- Various changes to canonicalise the order of things like
	  equal-priority results in the XML returned from
	  resolve_to_xml(), required so that the same regression-test
	  suite can be used with both MySQL and Oracle.
	- Add a test-script that invokes the actual "resolve" script.
	  The absence of this was responsible for 1.20 going out while
	  "resolve" itself was still not making its own $resolver
	  object, while the various test-scripts were.
	- It is now mandatory to pass a $resolver into
	  OpenURL::newFromCGI(), since memory problems will always
	  result if it is allowed to allocate its own.
	- Change API of OpenURL::newFromCGI() to reflect mandatory
	  $resolver and better match the default constructor, new().
	- Add Makefile.PL dependency on Scalar::Util.pm

1.21  Thu Apr 10 15:23:53 BST 2008
	- URGENT BUGFIX RELEASE: the resolve script now makes its own
	  $resolver object rather than allowing newFromCGI() to make
	  one on its behalf.  This ensures that the resolver stays
	  referenced (not only by the OpenURL's weak reference) until
	  the resolution is complete.

1.20  Thu Apr 10 12:36:28 BST 2008
	- Changes to memory management: since the Resolver, Database,
	  OpenURL and  ContextObject classes all contain references
	  back and forth, the next of mutually referencing objects was
	  not getting destructed when it went out of scope, leading to
	  memory exhaustion in long-running processes such as Apache
	  servers under heavy load.  That is now fixed through the use
	  of weak references back to the "owning" objects.  Tests show
	  that resolving 50,000 OpenURLs does not cause noticable
	  process growth.
	- Changes to regression-test suite to facilitate running when
	  not connected to network.
	- Added new test script, admin.t, testing the code that
	  underlies the web-based admin interface.
	- Refactored database-access code in Database.pm, greatly
	  reducing the number of similar functions and unifying
	  logging.
	- Add support for the Serial Alias object type, both in the
	  code and in the Admin web UI.  (Admin support for this is
	  not very useful yet, since, as with other kinds of link,
	  there is no way to maintain the links between an alias
	  object and the serial that it is an alias of.)
	- Added low-level object-API support for the service-to-serial
	  and genre-to-servicetype link tables.  (There's no point in
	  wiring them into the Admin UI, since they contain no
	  information other than link IDs.)
	- Search-result listing in Admin web UI does not offer links
	  to sort on virtual fields.
	- Search-result listing in Admin web UI restored to being
	  valid XHTML 1.0.
	- Top-level "resolve" script runs within a scoped block, to
	  ensure that all local ("my") variables go out of scope when
	  the script ends, whatever Apache's memory management regime
	  may be.  This contributes to fixing a memory-exhaustion
	  problem.

1.19  Wed Mar 26 17:22:50 GMT 2008
	- Use the DBI module's connect_cached() method to pool similar
	  connections.  This is important when running the resolver as
	  a web-service, since (currently) each invocation makes a new
	  Resolver object, which in turn makes its own Database
	  object, which until this change made its own DBI
	  connection.  Due to Perl's lazy garbage-collection, each
	  Apache process can accumulate a lot of these connections
	  before they get tidied up, and when (as is usually the case)
	  many Apache processes are used, they can exhaust the
	  back-end database's connection pool.
	- Log-levels (as specified for example by the opt_loglevel
	  query parameter) can now be specified symbolically, as a
	  comma-separated list of level names, as well as numerically.
	- New logging level, LIFECYCLE, notes the creation and
	  destruction of Resolver, OpenURL and Database objects.
	- The DBI logging level can now also be set, using the
	  opt_dbi_trace query parameter.
	- New function LogLevel::num() that converts a logging-level
	  string into a numeric value, complementing the existing
	  label().  Both these functions now handle composite values
	  (e.g. CONVERT01|LIFECYCLE).
	- New test-script for log-level utility functions.
	- Add brief documentation of logging (doc/logging).
	- Option setting is now always done via the option() method
	  rather than, as previously at Resolver creation time, via
	  direct assignment.  This ensures that the logging-related
	  special cases are handled uniformly.
	- Add static_log() function (not method) to the Resolver
	  class, for use in contexts where no Resolver object is
	  available to invoke the method on.
	- Copies are included (web/htdocs/1cate-test/) of the KEV
	  context-objects that were part of Openly's OpenURL 1.0 test
	  suite.  These will be useful one day for local testing.
	- More explicit diagnostics for some database problems.
	- Steps towards support for Oracle as the back-end database:
	  NOT YET COMPLETE.

1.18  Mon Feb 11 11:05:16 GMT 2008
	- Add the ability to delete resource database objects from the
	  Admin UI.

1.17  Thu Feb  7 16:26:37 CET 2008
	- Fix character-set handling AGAIN :-) The resolver library
	  was correcting accepting any specified character encoding
	  and correctly outputting UTF-8 with the necessary XML and
	  HTML specifications to sayt that's what it was doing, but
	  the top level resolver script was telling Perl to do UTF-8
	  encoding on output.  (Why didn't that cause problems when
	  running under Apache 1/mod_perl 1?  Because of the
	  character-set faries.)
	- Generalise rewrite rule so that /resolve/<dbname> URLs are

lib/Keystone/Resolver.pm
# $Id: Resolver.pm,v 1.36 2008-04-10 14:35:19 mike Exp $

package Keystone::Resolver;

use 5.008;
use strict;
use warnings;
use Keystone::Resolver::Utils;
use Keystone::Resolver::LogLevel;
use Keystone::Resolver::OpenURL;
use Keystone::Resolver::ContextObject;
use Keystone::Resolver::Descriptor;
use Keystone::Resolver::Database;
use Keystone::Resolver::ResultSet;

our $VERSION = '1.22';


=head1 NAME

Keystone::Resolver - an OpenURL resolver

=head1 SYNOPSIS

 use Keystone::Resolver;
 $resolver = new Keystone::Resolver();
 $openURL = $resolver->openURL($args, $base, $referer);
 ($type, $content) = $openURL->resolve();
 print "Content-type: $type\r\n\r\n$content";

=head1 DESCRIPTION

This is the top-level class of Index Data's Keystone Resolver.  It
delegates the process of resolving OpenURLs to a swarm of worker
classes.

=head1 METHODS

=cut

=head2 new()

 $resolver = new Keystone::Resolver();
 $resolver = new Keystone::Resolver(logprefix => "Keystone Resolver");
 $resolver = new Keystone::Resolver(logprefix => "Keystone Resolver",
                                    xsltdir => "/home/me/xslt");

Creates a new resolver that can be used to resolve OpenURLs.  If
arguments are provided, they are taken to be pairs that specify the
names and values of options.  See the documentation of the C<option()>
method below for information about specific options.

One option is special to this constructor: if C<_rw> is provided and
true, then the database is opened readwrite rather then readonly
(which is the default).

The resolver object accumulates some state as it goes along, so it
is generally more efficient to keep using a single resolver than to
make new one every time you need to resolve an OpenURL.

=cut

sub new {
    my $class = shift();
    my(%options) = @_;

    my $rw = delete $options{_rw};
    my $xsltdir = $ENV{KRxsltdir} || "../etc/xslt";
    my $this = bless {
	parser => undef,	# set when needed in parser()
	xslt => undef,		# set when needed in xslt()
	ua => undef,		# set when needed in ua()
	stylesheets => {},	# cache, populated by stylesheet()
	db => {},		# cache, populated by db()
	rw => $rw,
	options => {},
    }, $class;

    # Initial options can be overridden by creation-time arguments.
    # They should probably take default values from the Config table
    # of the RDB instead of hard-wired values.
    $this->option(logprefix => $0);
    $this->option(loglevel => $ENV{KRloglevel} || 0);
    $this->option(xsltdir => $xsltdir);
    foreach my $key (keys %options) {
	$this->option($key, $options{$key});
    }

    $this->log(Keystone::Resolver::LogLevel::LIFECYCLE, "new resolver $this");
    return $this;
}


sub DESTROY {
    my $this = shift();
    static_log(Keystone::Resolver::LogLevel::LIFECYCLE, "dead resolver $this");
    return; # The rest of this is unnecessary
    my @names = sort keys %{ $this->{db} };
    foreach my $name (@names) {
	static_log(Keystone::Resolver::LogLevel::LIFECYCLE,
		   "killing DB '$name'");
	undef $this->{db}->{$name};
    }
}


=head2 option()

 $level = $resolver->option("loglevel");
 $oldpath = $resolver->option(xsltdir => "/home/me/xslt");

Gets and sets options in a C<Keystone::Resolver> object.  When called with a
single argument, returns the value the resolver has for that key.
When called with two arguments, a key and a value, sets the specified
new value for that key and returns the old value anyway.

Supported options include:

=over 4

=item logprefix

The initial string emitted at the beginning of each line of debugging
output generated by the C<log()> method.  The default value is the
name of the running program.

=item loglevel

A bit mask indicating the categories of message that should be logged
by calls to the C<log()> method.  Should be set to the bitwise AND of
zero or more of the symbolic constants defined in
C<Keystone::Resolver::LogLevel>.  See the documentation of that module for a
description of the recognised categories.

=item xsltdir

The directory where the resolver looks for XSLT files.

=back

=cut

use vars qw($_last_loglevel $_last_logprefix);
# These must be set, probably by new(), before being used
$_last_loglevel = undef;
$_last_logprefix = undef;

# ### There is an issue with logging modality here: if a call is made
# with loglevel or dbi_trace > 0, then subsequent requests on the same
# resolver will inherit that logging level.  Maybe each request
# should explicitly zero the logging levels?
#
sub option {
    my $this = shift();
    my($key, $value) = @_;

    my $old = $this->{options}->{$key};
    if (defined $value) {
	# Special cases for "loglevel" to allow hex and octal bitmasks
	# and to parse non-numeric level-lists.
	if ($key eq "loglevel") {
	    $value = oct($value)
		if $value =~ /^0/;
	    $value = Keystone::Resolver::LogLevel::num($value)
		if $value !~ /^\d+$/;
	}
	#print STDERR "setting '$key' to '$value'\n";
	$this->{options}->{$key} = $value;

	# Save logging configuration for use of static_log()
	$_last_loglevel = $value if $key eq "loglevel";
	$_last_logprefix = $value if $key eq "logprefix";

	if ($key eq "dbi_trace") {
	    ### Two nastinesses here: the peek inside the database's
	    # internal structures, and the fact that we are operating
	    # on the default database.  We could "fix" the latter by
	    # changing the global state of the DBI library, but that
	    # would probably be even worse; or by getting db() from
	    # the OpenURL object (which might have a query parameter
	    # specifying which DB to work on) but we don't know what
	    # OpenURL object we're using.
	    $this->db()->{dbh}->trace($value);
	}

    }

    return $old;
}


=head2 log(), static_log()

 $resolver->log(Keystone::Resolver::LogLevel::CHITCHAT, "starting up");
 Keystone::Resolver::static_log(Keystone::Resolver::LogLevel::CHITCHAT, "end");

C<log()> Logs a message to the standard error stream if the log-level
of the resolver includes the level specified as the first argument in
its bitmask.  If so, the message consists of the logging prefix (by
default the name of the program), the label of the specified level in
parentheses, and all other arguments concatenated, finishing with a
newline.

C<static_log()> is provided for situtation in which no resolver object
is available, e.g. in C<DESTROY()> methods.  It behaves the same as
C<log()> but is a function, not a method.  Since it cannot consult the
options of a resolver object, it uses
I<the values most recently set into any resolver>.
For most applications, in which only a single resolver is in use, this
will work just fine.  Complex applications that use multiple resolvers
should not rely on the integrity of static logging.

=cut

sub log {
    my $this = shift();
    _log($this->option("loglevel"), $this->option("logprefix"), @_);
}


sub static_log {
    _log($_last_loglevel, $_last_logprefix, @_);
}


sub _log {
    my($loglevel, $logprefix, $level, @args) = @_;

    if ($loglevel & $level) {
	### could check another option for whether to include PID
	my $label = Keystone::Resolver::LogLevel::label($level);
	print STDERR "$logprefix ($label): ", @args, "\n";
	#use Carp; carp "$logprefix ($label): ", @args;
    }
}


=head2 openURL()

 $openURL = $resolver->openURL($args, $base, $referer);

Creates a new C<Keystone::Resolver::OpenURL> object using this
C<Keystone::Resolver> and the specified arguments and referer.  This
is a shortcut for

 new Keystone::Resolver::OpenURL($resolver, $args, $base, $referer)

=cut

sub openURL {
    my $this = shift();

    #use Carp qw(cluck); cluck("$$: creating new OpenURL(" . join(", ", map { defined $_ ? "'$_'" : "undef" } @_) . ")");
    return new Keystone::Resolver::OpenURL($this, @_);
}


=head2 parser()

 $parser = $resolver->parser();

Returns the XML parser associated with this resolver.  If it does not
yet have a parser, then one is created for it, cached for next time,
and returned.  The parser is an C<XML::LibXML> object: see the
documentation of that class for how to use it.

=cut

sub parser {
    my $this = shift();

    if (!defined $this->{parser}) {
	$this->{parser} = new XML::LibXML();
    }

    return $this->{parser};
}


=head2 xslt()

 $xslt = $resolver->xslt();

Returns the XSLT processor associated with this resolver.  If it does
not yet have a XSLT processor, then one is created for it, cached for
next time, and returned.  The XSLT processor is an C<XML::LibXSLT>
object: see the documentation of that class for how to use it.

=cut

sub xslt {
    my $this = shift();

    if (!defined $this->{xslt}) {
	$this->{xslt} = new XML::LibXSLT();
    }

    return $this->{xslt};
}


=head2 ua()

 $ua = $resolver->ua();

Returns the LWP User Agent associated with this resolver.  If it does
not yet have a User Agent, then one is created for it, cached for next
time, and returned.

=cut

sub ua {
    my $this = shift();

    if (!defined $this->{ua}) {
	$this->{ua} = new LWP::UserAgent();
    }

    return $this->{ua};
}


=head2 stylesheet()

 $stylesheet1 = $resolver->stylesheet();
 $stylesheet2 = $resolver->stylesheet("foo");

Returns a stylesheet object for the XSLT stylesheet named in the
argument, or for the default stylesheet if no argument is supplied.
The returned object is an <XML::LibXSLT::Stylesheet>: see the
documentation of that class for how to use it.

=cut

# $this->{stylesheets} is used only in this function.  It's a cache
# mapping a full stylesheet pathname to a duple consisting of that
# file's last modification time and the compiled stylesheet described
# by it.  The file is compiled if we're asked for it for the first
# time or if it's changed since the last compilation.
#
sub stylesheet {
    my $this = shift();
    my($name) = @_;

    $name ||= "default";
    my $cache = $this->{stylesheets};
    my $filename = $this->option("xsltdir") . "/$name.xsl";
    my(@stat) = stat($filename)
	or die "can't stat XSLT file '$filename': $!";
    my $mtime = $stat[9];
    $this->log(Keystone::Resolver::LogLevel::CACHECHECK,
	       "checking cache for XSLT file '$name', age $mtime");

    if (!defined $cache->{$name} ||
	$mtime > $cache->{$name}->[0]) {
	my $style_doc = $this->parser()->parse_file($filename);
	my $stylesheet = $this->xslt()->parse_stylesheet($style_doc);
	$cache->{$name} = [ $mtime, $stylesheet ];
	$this->log(Keystone::Resolver::LogLevel::PARSEXSLT,
		   "parsed XSLT file '$name', age $mtime");
    }

    return $cache->{$name}->[1];
}


=head2 db()

 $db = $resolver->db();
 $db = $resolver->db("kr-backup");

Returns the database object associated with this specified name for
this resolver.  If no name is provided, the default name specified by
the C<KRdb> environment variable is used; if this is also missing,
"kr" is used.  If the resolver does not yet have a
database handle associated with this name, then one is created for it,
cached for next time, and returned.  The handle is a
C<Keystone::Resolver::Database> object: see the documentation for how
to use it.

=cut

sub db {
    my $this = shift();
    my($name) = @_;

    $name ||= $ENV{KRdb} || "kr";
    my $cache = $this->{db};
    return $cache->{$name}
        if defined $cache->{$name};

    $cache->{$name} =
	new Keystone::Resolver::Database($this, $name, $this->{rw});
    ### We want the cached Database references to be weak, so that
    #   the databases get destroyed before the resolver that they
    #   depend on.  Weakening the reference should do this but doesn't
    #	seem to have any effect (suggesting a bug in Perl?).  So we
    #	won't do it, in case it has unanticipated side-effects.
    #Scalar::Util::weaken($cache->{$name});
    return $cache->{$name};
}


=head1 AUTHOR

Mike Taylor E<lt>mike@indexdata.comE<gt>

First version Tuesday 9th March 2004.

=head1 SEE ALSO

C<Keystone::Resolver::OpenURL>,
C<Keystone::Resolver::Result>,
C<Keystone::Resolver::LogLevel>,
C<Keystone::Resolver::ContextObject>,
C<Keystone::Resolver::Database>,
C<Keystone::Resolver::Descriptor>,
C<Keystone::Resolver::Test>.

=cut


1;
